onnx/defs/math/defs.cc

```diff
@@ -1666,6 +1667,301 @@ ONNX_OPERATOR_SET_SCHEMA(
+            c = target[n][d_1]
+            loss[n][d_1] = -input[n][c][d_1]
+
+    # print(loss)
+    # [[-3. -2.]
+    #  [-0. -2.]]
+
+Example 2:
+
+    # weighted negative log likelihood loss, sum reduction
+    N, C, d1 = 2, 3, 2
+    input = [[[1.0, 2.0], [2.0, 2.0], [3.0, 2.0]], 
+            [[0.0, 1.0], [2.0, 2.0], [1.0, 2]]]
+    target = [[2, 1], [0, 2]]
+    weight = [0.2, 0.3, 0.1]
+    loss = np.zeros((N, d1))
+    for n in range(N):
+        for d_1 in range(d1):
+            c = target[n][d_1]
+            loss[n][d_1] = -input[n][c][d_1] * weight[c]
+
+    loss = np.sum(loss)
+    # print(loss)
+    # -1.1
+
+Example 3:
+
+    # weighted negative log likelihood loss, mean reduction
+    N, C, d1 = 2, 3, 2
+    input = [[[1.0, 2.0], [2.0, 2.0], [3.0, 2.0]], 
+            [[0.0, 1.0], [2.0, 2.0], [1.0, 2]]]
+    target = [[2, 1], [0, 2]]
+    weight = [0.2, 0.3, 0.1]
+    loss = np.zeros((N, d1))
+    weight_total = 0
+    for n in range(N):
+        for d_1 in range(d1):
+            c = target[n][d_1]
+            loss[n][d_1] = -input[n][c][d_1] * weight[c]
+            weight_total = weight_total + weight[c]
+
+    loss = np.sum(loss) / weight_total
+    # print(loss)
+    # -1.57
+)DOC";
+
+TensorProto ToDimensionOneTensor(int32_t value) {
+  auto t = ToTensor(std::vector<int32_t>({value}));
+  t.add_dims(1);
+  return t;
+}
+
+ONNX_OPERATOR_SET_SCHEMA(
+    NllLoss,
+    12,
+    OpSchema()
+        .SetDoc(NllLoss_ver12_doc)
+        .Input(
+            0,
+            "input",
+            "Input tensor of shape (N, C) or (N, C, d1, d2, ..., dk).",
+            "T")
+        .Input(
+            1,
+            "target",
+            "Target tensor of shape (N) or (N, d1, d2, ..., dk). Target element value shall be in range of [0, C).",
+            "Tind")
+        .Input(
+            2,
+            "weight",
+            "Optional rescaling weight tensor. "
+            "If given, it has to be a tensor of size C. Otherwise, it is treated as if having all ones.",
+            "T",
+            OpSchema::Optional)
+        .Output(0, "loss", "The negative log likelihood loss", "T")
+        .Attr(
+            "reduction",
+            "Type of reduction to apply to loss: none, sum, mean(default). "
+            "'none': the output is the loss for each sample in the batch."
+            "'sum': the output will be summed. "
+            "'mean': the sum of the output will be divided by the batch_size.",
+            AttributeProto::STRING,
+            std::string("mean"))
+        .TypeConstraint(
+            "T",
+            {"tensor(float16)", "tensor(float)", "tensor(double)"},
+            "Constrain input, weight, and output types to floating-point tensors.")
+        .TypeConstraint(
+            "Tind",
+            {"tensor(int32)", "tensor(int64)"},
+            "Constrain target to integer types")
+        .AddQueriedFunctionBody([](FunctionBodyQueryContext& ctx) { // no weight, reduction is "none"
```

Comment:
>AddQueriedFunctionBody [](start = 9, length = 22)

general speaking, I feel that this approach is not extensible... 

In this op, there is 6 combinations....
What about other ops if the options are combinatorial.... 

Is this the designed way to handle conditional function body?

Reply:
@gramalingam, @linkerzhang might have some opinions.

Reply:
My suggestion (which is up earlier) is to register a single function that takes the Context as input and returns the function-body.

Reply:
If the options are combinatorial, then there will be combinatorial number of graph bodies. I do not see how it can be reduced in the current framework. Maybe it will be solved with MLIR dialect?

Reply:
@liqunfu : if we define a single function that takes the Context as input and returns the function-body, it will look like below:
```
   NodeSet nodes;
   nodes.add(common_prefix_nodes);
   if (condition1(context)) nodes.add(variant1_A); else nodes.add(variant1_B);
   nodes.add(common_middle_nodes);
   if (condition2(context)) nodes.add(variant2_A); else nodes.add(variant2_B);
   nodes.add(common_suffix);
```
If this is converted to the existing approach, we will have to map every execution-path in the above code to a different (condition,body) pair, leading to the combinatorial explosion that @SherlockNoMad is talking about.

Reply:
@gramalingam this sounds very good. Let me give it a try.

Reply:
updated the PR according to @gramalingam 's suggestion.